### PR TITLE
Performance improvements for `MultivariatePolynomialFunction` and `PolynomialPathFitter` 

### PR DESCRIPTION
**Pull request by nickbianco**
Fixes issue #3759, #3762. Partially addresses #3761.

### Brief summary of changes

This PR implements several performance improvements related to `FunctionBasedPath` by speeding up function evaluations in `MultivariatePolynomialFunction` and adding a few new options to `PolynomialPathFitter`.  A brief list of the main changes:

- Added the internal helper class `MultivariatePolynomial` for computing basic symbolic polynomial function operations. 
- Refactored `MultivariatePolynomialFunction` to compute function values and derivatives based on Horner's scheme.
- Added the method `generateDerivativeFunction` and `generatePartialVelocityFunction` to `MultivariatePolynomialFunction` for conveniently generating moment arm and lengthening speed functions from a muscle path function.
- Added the properties `include_moment_arm_functions` and `include_lengthening_speed_function` to `PolynomialPathFitter`, which leverage the above methods.
- Replaced current least-squares fitting in `PolynomialPathFitter` with `SimTK::FactorQTZ`, which should be more robust to poorly conditioned `A` matrices (this is often the case for the multivariate polynomials since the terms can be highly correlated).
- Fixed the bug described in #3759. 

### Testing I've completed

Added related tests to `testFunctions.cpp` and `testWrapping.cpp`. A summery of the performance benchmarking is below (seeing a ~20-25% speed up).

### CHANGELOG.md (choose one)

- updated.

[perf-win]

### Performance analysis

Platform: Windows, self-hosted runner

|                           Test Name | lhs [secs] | stderr [secs] | rhs [secs] | stderr [secs] | Speedup |
| ----------------------------------- | ---------- | ------------- | ---------- | ------------- | ------- |
|                               Arm26 |       0.36 |          0.00 |       0.36 |          0.00 |    1.00 |
|                            Gait2354 |       0.44 |          0.00 |       0.42 |          0.00 |    1.03 |
|                     MocoSlidingMass |       1.51 |          0.00 |       1.52 |          0.00 |    1.00 |
|                     passive_dynamic |       5.63 |          0.00 |       5.79 |          0.00 |    0.97 |
|          passive_dynamic_noanalysis |       3.49 |          0.00 |       3.48 |          0.00 |    1.00 |
|                      RajagopalModel |       9.06 |          0.00 |       9.10 |          0.01 |    0.99 |
| ToyDropLanding_function_based_paths |      15.92 |          1.35 |      12.63 |          0.00 |    1.26 |
|                      ToyDropLanding |      13.68 |          0.01 |      13.72 |          0.01 |    1.00 |
|            ToyDropLanding_nomuscles |       0.57 |          0.00 |       0.57 |          0.00 |    1.01 |

- - -
This change is [Reviewable](https://reviewable.io/reviews/opensim-org/opensim-core/3767)

### CODE REVIEW

**Comment by nickbianco** [comment]
@pepbos would you be interested in reviewing this? It is related to muscle wrapping and performance.

File: OpenSim/Common/MultivariatePolynomialFunction.cpp [hunk 1]
```diff
@@ -17,205 +17,362 @@
...
-
 #include "Exception.h"
 
 using namespace OpenSim;
 
+/** 
+ * A helper class to construct and manipulate multivariate polynomials using 
+ * symbolic operations.
+ */ 
+typedef std::map<std::string, int> Term;
```
[the remark this answers is not in the document]
Comment (pepbos) [reply OpenSim/Common/MultivariatePolynomialFunction.cpp:28]: The name "Term" is a bit confusing. Perhaps use "Monomial" or "PowerProduct" if that is indeed what it is.

File: OpenSim/Common/MultivariatePolynomialFunction.cpp [hunk 2]
```diff
@@ -17,205 +17,362 @@
...
+    // CALC METHODS
+    /**
+     * Calculate the order of the polynomial.
+     */
+    int calcOrder() const {
+        int order = 0;
+        for (auto it = begin(); it != end(); it++) {
+            int sum = 0;
+            for (auto it2 = it->first.begin(); it2 != it->first.end(); it2++) {
+                sum += it2->second;
```
Comment (pepbos) [inline OpenSim/Common/MultivariatePolynomialFunction.cpp:114]: This loop, and the ones below it, could really use short comments to describe what is going on.

File: OpenSim/Common/MultivariatePolynomialFunction.cpp [hunk 3]
```diff
@@ -17,205 +17,362 @@
...
+        MultivariatePolynomial deriv;
+        for (auto it = poly.begin(); it != poly.end(); it++) {
+            Term t = it->first;
+            double c = it->second;
+            for (auto it2 = t.begin(); it2 != t.end(); it2++) {
+                Term t2 = t;
+                if (it2->first == var) {
+                    t2.erase(it2->first);
+                    if (it2->second > 1) {
+                        t2[it2->first] = it2->second - 1;
```
Comment (pepbos) [inline OpenSim/Common/MultivariatePolynomialFunction.cpp:166]: It would be really great to add an accessing member function that clarifies what we get, e.g.:
`double getCoefficient(const PowerProduct& key)`
and similarly for the map of variable to power.
Because we have two nested maps, i really have to read carefully to understand which one i am accessing, the coefficient, or the power.

**Review by pepbos — CHANGES_REQUESTED** [review]
Very nice! The `calcPolynomialHelper` was impressive, but hard to follow :) So, this is a great improvement.

I do think that the code could benefit a lot from short comments in between the math, such that an outsider understands it easily.
Similarly for the nested maps in `MultivariatePolynomialFunction` a named accessing function would help in understanding what I am accessing, as opposed to using the `[]` operator.

There are also some trailing spaces here and there.

**Review by nickbianco — COMMENTED** [review]
Thanks for the great review @pepbos! I've addressed your comments plus a few other small things I noticed along the way.

> *[Reviewable](https://reviewable.io/reviews/opensim-org/opensim-core/3767#-:-NwQV37K4H6eiE0VL_R3:bl7uq43)* status: 0 of 9 files reviewed, 7 unresolved discussions (waiting on @pepbos)

___
*[`OpenSim/Common/MultivariatePolynomialFunction.cpp` line 0 at r2](https://reviewable.io/reviews/opensim-org/opensim-core/3767#gh-1581054292:-NwQNBnd1BJrY9foDbJg:b-896fix) ([raw file](https://github.com/opensim-org/opensim-core/blob/cc18bf4a573755a88071e98fd61cfb18c71b1f81/OpenSim/Common/MultivariatePolynomialFunction.cpp#L0)):*
<details><summary><i>Previously, pepbos (pepbos) wrote…</i></summary><blockquote>

*(Reviewable was [unable](https://github.com/Reviewable/Reviewable/issues/14) to map [this GitHub inline comment thread](https://github.com/opensim-org/opensim-core/pull/3767#discussion_r1581054292) to the right spot — sorry!)*

nice cleanup!
</blockquote></details>

Done.
___
*[`OpenSim/Common/MultivariatePolynomialFunction.cpp` line 28 at r2](https://reviewable.io/reviews/opensim-org/opensim-core/3767#gh-1580958438:-NwQNClx7ey8ZuonZfq9:b-896fix) ([raw file](https://github.com/opensim-org/opensim-core/blob/cc18bf4a573755a88071e98fd61cfb18c71b1f81/OpenSim/Common/MultivariatePolynomialFunction.cpp#L28)):*
<details><summary><i>Previously, pepbos (pepbos) wrote…</i></summary><blockquote>

The name "Term" is a bit confusing. Perhaps use "Monomial" or "PowerProduct" if that is indeed what it is.
</blockquote></details>

Done.
___
*[`OpenSim/Common/MultivariatePolynomialFunction.cpp` line 34 at r2](https://reviewable.io/reviews/opensim-org/opensim-core/3767#gh-1580959778:-NwQND_N3QwB75FMDRXR:b-896fix) ([raw file](https://github.com/opensim-org/opensim-core/blob/cc18bf4a573755a88071e98fd61cfb18c71b1f81/OpenSim/Common/MultivariatePolynomialFunction.cpp#L34)):*
<details><summary><i>Previously, pepbos (pepbos) wrote…</i></summary><blockquote>

trailing space
</blockquote></details>

Done.
___
*[`OpenSim/Common/MultivariatePolynomialFunction.cpp` line 34 at r2](https://reviewable.io/reviews/opensim-org/opensim-core/3767#gh-1581011018:-NwQNHtb9Znvl70v4hk2:bgfwod4) ([raw file](https://github.com/opensim-org/opensim-core/blob/cc18bf4a573755a88071e98fd61cfb18c71b1f81/OpenSim/Common/MultivariatePolynomialFunction.cpp#L34)):*
<details><summary><i>Previously, pepbos (pepbos) wrote…</i></summary><blockquote>

Ok, I think I get it.

If I have a polynomial: $f = \pi * x^2 * y + 0.5 * x^3 * y^{23}$
this will be stored in a map as:

- First element:
    - key: a `Term` with key-value pairs: { ("x", 2), ("y", 1) }
    - value: $\pi$
- Second element:
    - key: a `Term` with key-value pairs: {("x", 3), ("y", 23)}
    - value: $0.5$

Is that correct?

</blockquote></details>

Yes, that's correct. I've added some clarifying comments.
___
*[`OpenSim/Common/MultivariatePolynomialFunction.cpp` line 114 at r2](https://reviewable.io/reviews/opensim-org/opensim-core/3767#gh-1581090437:-NwQRNOf4xC0OiIdxmP9:b-896fix) ([raw file](https://github.com/opensim-org/opensim-core/blob/cc18bf4a573755a88071e98fd61cfb18c71b1f81/OpenSim/Common/MultivariatePolynomialFunction.cpp#L114)):*
<details><summary><i>Previously, pepbos (pepbos) wrote…</i></summary><blockquote>

This loop, and the ones below it, could really use short comments to describe what is going on.
</blockquote></details>

Done.
___
*[`OpenSim/Common/MultivariatePolynomialFunction.cpp` line 166 at r2](https://reviewable.io/reviews/opensim-org/opensim-core/3767#gh-1581094175:-NwQUA1i0Kn0V1bSMif0:b-896fix) ([raw file](https://github.com/opensim-org/opensim-core/blob/cc18bf4a573755a88071e98fd61cfb18c71b1f81/OpenSim/Common/MultivariatePolynomialFunction.cpp#L166)):*
<details><summary><i>Previously, pepbos (pepbos) wrote…</i></summary><blockquote>

It would be really great to add an accessing member function that clarifies what we get, e.g.:
`double getCoefficient(const PowerProduct& key)`
and similarly for the map of variable to power.
Because we have two nested maps, i really have to read carefully to understand which one i am accessing, the coefficient, or the power.
</blockquote></details>

Done.
___
*[`OpenSim/Common/MultivariatePolynomialFunction.cpp` line 189 at r2](https://reviewable.io/reviews/opensim-org/opensim-core/3767#gh-1581056741:-NwQRReR9by3MfJhWq1I:b-896fix) ([raw file](https://github.com/opensim-org/opensim-core/blob/cc18bf4a573755a88071e98fd61cfb18c71b1f81/OpenSim/Common/MultivariatePolynomialFunction.cpp#L189)):*
<details><summary><i>Previously, pepbos (pepbos) wrote…</i></summary><blockquote>

replace `if(term.find(var) == term.end()) { ... }` by `term[var] += exponent`?
I think this should insert the key with value zero, if there was none before
</blockquote></details>

Done.

**Review by nickbianco — COMMENTED** [review]
Thanks @pepbos! 

Just one minor note related to renaming methods, but I think we're good to go. 

> *[Reviewable](https://reviewable.io/reviews/opensim-org/opensim-core/3767#-:-NweXZvj3MCxeIr5rTch:b-dvnm0)* status: 0 of 9 files reviewed, 14 unresolved discussions (waiting on @pepbos)

___
*[`OpenSim/Common/MultivariatePolynomialFunction.cpp` line 142 at r3](https://reviewable.io/reviews/opensim-org/opensim-core/3767#gh-1583174512:-NweUP2C2zFQiK-xuEXK:bb2v9yb) ([raw file](https://github.com/opensim-org/opensim-core/blob/ef75fbc9491f099787dd24264acbc72ce21b944e/OpenSim/Common/MultivariatePolynomialFunction.cpp#L142)):*
<details><summary><i>Previously, pepbos (pepbos) wrote…</i></summary><blockquote>

`static`
</blockquote></details>

This method can't be `static`: it is searching/accessing the current polynomial (i.e., `this`).
___
*[`OpenSim/Common/MultivariatePolynomialFunction.cpp` line 154 at r3](https://reviewable.io/reviews/opensim-org/opensim-core/3767#gh-1583175062:-NweVL8t0-KpJq8WUOo8:b-896fix) ([raw file](https://github.com/opensim-org/opensim-core/blob/ef75fbc9491f099787dd24264acbc72ce21b944e/OpenSim/Common/MultivariatePolynomialFunction.cpp#L154)):*
<details><summary><i>Previously, pepbos (pepbos) wrote…</i></summary><blockquote>

This method is a bit expensive, because you already have the key-value pair in the cases below where you use it. Perhaps it can be removed, because with the `Monomial` introduction, it is much more clear to me that you will iterate over variable-exponent pairs.
See comments below where used
</blockquote></details>

Done.
___
*[`OpenSim/Common/MultivariatePolynomialFunction.cpp` line 175 at r3](https://reviewable.io/reviews/opensim-org/opensim-core/3767#gh-1583199603:-NweUiQB2G78mHHQMTLB:b-896fix) ([raw file](https://github.com/opensim-org/opensim-core/blob/ef75fbc9491f099787dd24264acbc72ce21b944e/OpenSim/Common/MultivariatePolynomialFunction.cpp#L175)):*
<details><summary><i>Previously, pepbos (pepbos) wrote…</i></summary><blockquote>

`int exponent = itv->second;`
</blockquote></details>

Done.
___
*[`OpenSim/Common/MultivariatePolynomialFunction.cpp` line 242 at r3](https://reviewable.io/reviews/opensim-org/opensim-core/3767#gh-1583200367:-NweUmhg1diCoyfvGDBO:b-896fix) ([raw file](https://github.com/opensim-org/opensim-core/blob/ef75fbc9491f099787dd24264acbc72ce21b944e/OpenSim/Common/MultivariatePolynomialFunction.cpp#L242)):*
<details><summary><i>Previously, pepbos (pepbos) wrote…</i></summary><blockquote>

`int exponent = itm->second`
</blockquote></details>

Done.
___
*[`OpenSim/Common/MultivariatePolynomialFunction.cpp` line 406 at r3](https://reviewable.io/reviews/opensim-org/opensim-core/3767#gh-1583207077:-NweUxjM6JWhUEz98NSS:b-896fix) ([raw file](https://github.com/opensim-org/opensim-core/blob/ef75fbc9491f099787dd24264acbc72ce21b944e/OpenSim/Common/MultivariatePolynomialFunction.cpp#L406)):*
<details><summary><i>Previously, pepbos (pepbos) wrote…</i></summary><blockquote>

`Monomial monomial;`
</blockquote></details>

Done.
___
*[`OpenSim/Common/MultivariatePolynomialFunction.cpp` line 434 at r3](https://reviewable.io/reviews/opensim-org/opensim-core/3767#gh-1583233936:-NweWL7C-W88D-JLdv_3:b-nfdho) ([raw file](https://github.com/opensim-org/opensim-core/blob/ef75fbc9491f099787dd24264acbc72ce21b944e/OpenSim/Common/MultivariatePolynomialFunction.cpp#L434)):*
<details><summary><i>Previously, pepbos (pepbos) wrote…</i></summary><blockquote>

`calcMonomialValues`
</blockquote></details>

I'm fine updating this (and `calcMonomialDerivatives`), but it does make these method names inconsistent with the public methods `getTermValues` and `getTermDerivatives` in `MultivariatePolynomialFunction`. I could update these too, but it would be an API change. `getTermValues` and `getTermDerivatives` were only recently introduced (by me) for use in `PolynomialPathFitter`,  but if users of OpenSim 4.5 start using these methods then an API change would be problematic.
___
*[`OpenSim/Common/MultivariatePolynomialFunction.cpp` line 447 at r3](https://reviewable.io/reviews/opensim-org/opensim-core/3767#gh-1583233383:-NweX2xO6zOt-yBps4si:b-otilcs) ([raw file](https://github.com/opensim-org/opensim-core/blob/ef75fbc9491f099787dd24264acbc72ce21b944e/OpenSim/Common/MultivariatePolynomialFunction.cpp#L447)):*
<details><summary><i>Previously, pepbos (pepbos) wrote…</i></summary><blockquote>

"calcMonomialDerivatives`
</blockquote></details>

See comment above.
